README.md
# ![chartjs-gauge logo](./samples/logo.svg) chartjs-gauge

Simple gauge chart for [Chart.js](https://www.chartjs.org/)

## Samples

- [Gauge Chart](https://codepen.io/haiiaaa/pen/rNVbmYy)
- [Gauge Chart with datalabels plugin](https://codepen.io/haiiaaa/pen/KKpYmRz)
- [Gauge Chart with datalabels plugin displaying labels](https://codepen.io/haiiaaa/pen/qBdwmyY)

## Install

- **yarn** install: `yarn add chart.js chartjs-gauge`
- **npm** install: `npm install --save chart.js chartjs-gauge`

<!---
## Documentation

- [Samples](https://codepen.io/???/)
--->
## Configuration Options

The gauge chart is based on the [Doughnut](https://www.chartjs.org/docs/latest/charts/doughnut.html#dataset-properties) type. It  defines the following additional configuration options. These options are merged with the global chart configuration options, `Chart.defaults.global`, to form the options passed to the chart.

| Name | Type | Default | Description
| ---- | ---- | ------- | -----------
| `needle.radiusPercentage`           | `number`                                                           | `2`                        | Needle circle radius as the percentage of the chart area width.
| `needle.widthPercentage`            | `number`                                                           | `3.2`                      | Needle width as the percentage of the chart area width.
| `needle.lengthPercentage`           | `number`                                                           | `80`                       | Needle length as the percentage of the interval between inner radius (0%) and outer radius (100%) of the arc.
| `needle.color`                      | [`Color`](https://www.chartjs.org/docs/latest/general/colors.html) | `'rgba(0, 0, 0, 1)'`       | The color of the needle.
| `valueLabel.display`                | `boolean`                                                          | `true`                     | If true, display the value label.
| `valueLabel.formatter`              | `function`                                                         | `Math.round`               | Returns the string representation of the value as it should be displayed on the chart.
| `valueLabel.fontSize`               | `number`                                                           | `undefined`                | The font size of the label.
| `valueLabel.color`                  | [`Color`](https://www.chartjs.org/docs/latest/general/colors.html) | `'rgba(255, 255, 255, 1)'` | The text color of the label.
| `valueLabel.backgroundColor`        | [`Color`](https://www.chartjs.org/docs/latest/general/colors.html) | `'rgba(0, 0, 0, 1)'`       | The background color of the label.
| `valueLabel.borderRadius`           | `number`                                                           | `5`                        | Border radius of the label.
| `valueLabel.padding.top`            | `number`                                                           | `5`                        | Top padding of the label.
| `valueLabel.padding.right`          | `number`                                                           | `5`                        | Right padding of the label.
| `valueLabel.padding.bottom`         | `number`                                                           | `5`                        | Bottom padding of the label.
| `valueLabel.padding.left`           | `number`                                                           | `5`                        | Left padding of the label.
| `valueLabel.bottomMarginPercentage` | `number`                                                           | `5`                        | Bottom margin as the percentage of the chart area width.


## Default Options

It is common to want to apply a configuration setting to all created gauge charts. The global gauge chart settings are stored in `Chart.defaults.gauge`. Changing the global options only affects charts created after the change. Existing charts are not changed.

For example, to configure all line charts with `radiusPercentage = 5` you would do:
```javascript
Chart.defaults.gauge.needle.radiusPercentage = 5;
```

## Dataset Properties

The gauge chart requires a value to be specified for the dataset. This is used to draw the needle for the dataset.

| Name       | Type     | Default     | Description
| ---------- | -------- | ----------- | -----------
| `value`    | `number` | `undefined` | Value used for the needle.
| `minValue` | `number` | `0`         | Used to offset the start value.


## Example

```javascript
var ctx = document.getElementById("canvas").getContext("2d");

var chart = new Chart(ctx, {
  type: 'gauge',
  data: {
    datasets: [{
      value: 0.5,
      minValue: 0,
      data: [1, 2, 3, 4],
      backgroundColor: ['green', 'yellow', 'orange', 'red'],
    }]
  },
  options: {
    needle: {
      radiusPercentage: 2,
      widthPercentage: 3.2,
      lengthPercentage: 80,
      color: 'rgba(0, 0, 0, 1)'
    },
    valueLabel: {
      display: true,
      formatter: (value) => {
        return '$' + Math.round(value);
      },
      color: 'rgba(255, 255, 255, 1)',
      backgroundColor: 'rgba(0, 0, 0, 1)',
      borderRadius: 5,
      padding: {
        top: 10,
        bottom: 10
      }
    }
  }
});
```

## License

chartjs-gauge is available under the [MIT license](https://opensource.org/licenses/MIT).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.3.0] - 2021-05-05

### Added

- [#18](https://github.com/haiiaaa/chartjs-gauge/issues/18) Not starting at zero

## [0.2.0] - 2020-06-04

### Added

- [#5](https://github.com/haiiaaa/chartjs-gauge/issues/5) Change valuelabel font size
- Changelog

## [0.1.2] - 2020-04-23

### Fixed

- [#4](https://github.com/haiiaaa/chartjs-gauge/issues/4) TextHeight calculation can be improved

## [0.1.1] - 2020-04-02

### Fixed

- [#3](https://github.com/haiiaaa/chartjs-gauge/issues/3) Support IE11
- [#1](https://github.com/haiiaaa/chartjs-gauge/issues/1) rotation doesn't work properly

## [0.1.0] - 2019-11-14

### Added

- Initial version

[unreleased]: https://github.com/haiiaaa/chartjs-gauge/compare/release/0.3.0...master
[0.3.0]: https://github.com/haiiaaa/chartjs-gauge/compare/release/0.2.0...release/0.3.0
[0.2.0]: https://github.com/haiiaaa/chartjs-gauge/compare/release/0.1.2...release/0.2.0
[0.1.2]: https://github.com/haiiaaa/chartjs-gauge/compare/release/0.1.1...release/0.1.2
[0.1.1]: https://github.com/haiiaaa/chartjs-gauge/compare/release/0.1.0...release/0.1.1
[0.1.0]: https://github.com/haiiaaa/chartjs-gauge/compare/9e290ff9a93ca5ff7f7e312cf1ddbbc1831e84a9...release/0.1.0


test/specs/controller.gauge.tests.js
describe('Chart.controllers.gauge', () => {
  it('should be registered as dataset controller', () => {
    expect(typeof Chart.controllers.gauge).toBe('function');
  });

  it('should be constructed', () => {
    const chart = window.acquireChart({
      type: 'gauge',
      data: {
        datasets: [{
          value: 1,
          data: [],
        }],
      },
    });

    const meta = chart.getDatasetMeta(0);
    expect(meta.type).toBe('gauge');
    expect(meta.controller).not.toBe(undefined);
    expect(meta.controller.index).toBe(0);
    expect(meta.data).toEqual([]);

    meta.controller.updateIndex(1);
    expect(meta.controller.index).toBe(1);
  });

  it('should create gauge arc elements for each data item during initialization', () => {
    const chart = window.acquireChart({
      type: 'gauge',
      data: {
        datasets: [{
          value: 2,
          data: [10, 15, 0, 4],
        }],
      },
    });

    const meta = chart.getDatasetMeta(0);
    expect(meta.data.length).toBe(4); // 4 rectangles created
    expect(meta.data[0] instanceof Chart.elements.Arc).toBe(true);
    expect(meta.data[1] instanceof Chart.elements.Arc).toBe(true);
    expect(meta.data[2] instanceof Chart.elements.Arc).toBe(true);
    expect(meta.data[3] instanceof Chart.elements.Arc).toBe(true);
  });

  it('should scale values', () => {
    const chart = window.acquireChart({
      type: 'gauge',
      data: {
        datasets: [{
          value: 1,
          data: [2, 3, 4],
        }],
      },
      options: {
        rotation: -Math.PI,
        circumference: Math.PI,
      },
    });

    const meta = chart.getDatasetMeta(0);

    expect(meta.data.length).toBe(3);

    // Only startAngle, endAngle and circumference should be different.
    [
      { c: Math.PI / 2, s: -Math.PI, e: -Math.PI / 2 },
      { c: Math.PI / 4, s: -Math.PI / 2, e: -Math.PI / 4 },
      { c: Math.PI / 4, s: -Math.PI / 4, e: 0 },
    ].forEach((expected, i) => {
      expect(meta.data[i]._model.circumference).toBeCloseTo(expected.c, 8, `circumference ${i}`);
      expect(meta.data[i]._model.startAngle).toBeCloseTo(expected.s, 8, `startAngle ${i}`);
      expect(meta.data[i]._model.endAngle).toBeCloseTo(expected.e, 8, `endAngle ${i}`);
    });
  });

  it('should scale values for minValue', () => {
    const chart = window.acquireChart({
      type: 'gauge',
      data: {
        datasets: [{
          value: 0,
          minValue: -1, // number | true
          data: [1, 2, 3], // or [0, 2, 3, 4] with [null, 'green', 'yellow', 'red']
        }],
      },
      options: {
        rotation: -Math.PI,
        circumference: Math.PI,
      },
    });

    const meta = chart.getDatasetMeta(0);

    expect(meta.data.length).toBe(3);

    // Only startAngle, endAngle and circumference should be different.
    [
      { c: Math.PI / 2, s: -Math.PI, e: -Math.PI / 2 },
      { c: Math.PI / 4, s: -Math.PI / 2, e: -Math.PI / 4 },
      { c: Math.PI / 4, s: -Math.PI / 4, e: 0 },
    ].forEach((expected, i) => {
      expect(meta.data[i]._model.circumference).toBeCloseTo(expected.c, 8, `circumference ${i}`);
      expect(meta.data[i]._model.startAngle).toBeCloseTo(expected.s, 8, `startAngle ${i}`);
      expect(meta.data[i]._model.endAngle).toBeCloseTo(expected.e, 8, `endAngle ${i}`);
    });
  });
});


karma.conf.js
const commonjs = require('rollup-plugin-commonjs'); // eslint-disable-line import/no-extraneous-dependencies
const resolve = require('rollup-plugin-node-resolve'); // eslint-disable-line import/no-extraneous-dependencies
const istanbul = require('rollup-plugin-istanbul'); // eslint-disable-line import/no-extraneous-dependencies
const builds = require('./rollup.config');

module.exports = (karma) => {
  const args = karma.args || {};

  // Use the same rollup config as our dist files: when debugging (--watch),
  // we will prefer the unminified build which is easier to browse and works
  // better with source mapping. In other cases, pick the minified build to
  // make sure that the minification process (terser) doesn't break anything.
  const regex = args.watch ? /chartjs-gauge\.js$/ : /chartjs-gauge\.min\.js$/;
  const build = builds.filter(v => v.output.file.match(regex))[0];

  if (args.watch) {
    build.output.sourcemap = 'inline';
  }

  karma.set({
    frameworks: ['jasmine'],
    reporters: ['progress'],
    browsers: ['chrome', 'firefox'],
    // browsers: ['ChromeHeadless'],
    logLevel: karma.LOG_WARN,

    // Explicitly disable hardware acceleration to make image
    // diff more stable when ran on Travis and dev machine.
    // https://github.com/chartjs/Chart.js/pull/5629
    customLaunchers: {
      chrome: {
        base: 'Chrome',
        flags: [
          '--disable-accelerated-2d-canvas',
        ],
      },
      firefox: {
        base: 'Firefox',
        prefs: {
          'layers.acceleration.disabled': true,
        },
      },
    },

    files: [
      { pattern: 'node_modules/chart.js/dist/Chart.bundle.min.js', watched: false },
      { pattern: 'modules/chart.js/test/index.js', watched: false },
      'src/index.js', // watch throws
    ].concat(args.inputs || ['test/specs/**/*.js']),

    preprocessors: {
      'modules/chart.js/test/index.js': ['rollup'],
      'src/index.js': ['sources'],
    },

    rollupPreprocessor: {
      plugins: [
        resolve(),
        commonjs(),
      ],
      output: {
        name: 'test',
        format: 'umd',
      },
    },

    customPreprocessors: {
      sources: {
        base: 'rollup',
        options: build,
      },
    },

    // These settings deal with browser disconnects. We had seen test flakiness from Firefox
    // [Firefox 56.0.0 (Linux 0.0.0)]: Disconnected (1 times), because no message in 10000 ms.
    // https://github.com/jasmine/jasmine/issues/1327#issuecomment-332939551
    browserDisconnectTolerance: 3,
  });

  // https://swizec.com/blog/how-to-run-javascript-tests-in-chrome-on-travis/swizec/6647
  if (process.env.TRAVIS) {
    karma.customLaunchers.chrome.flags.push('--no-sandbox');
  }

  if (args.coverage) {
    karma.reporters.push('coverage');
    karma.coverageReporter = {
      dir: 'coverage/',
      reporters: [
        { type: 'html', subdir: 'html' },
        { type: 'lcovonly', subdir: '.' },
      ],
    };
    [
      karma.rollupPreprocessor,
      karma.customPreprocessors.sources.options,
    ].forEach(v => {
      (v.plugins || (v.plugins = [])).unshift(
        istanbul({
          include: 'src/**/*.js',
        }),
      );
    });
  }
};


rollup.config.js
const babel = require('rollup-plugin-babel');
const { terser } = require('rollup-plugin-terser');
const pkg = require('./package.json');

const input = 'src/index.js';
const banner = `/*!
 * chartjs-gauge.js v${pkg.version}
 * ${pkg.homepage}
 * (c) ${new Date().getFullYear()} chartjs-gauge.js Contributors
 * Released under the MIT License
 */`;

module.exports = [
  // UMD builds (excluding Chart)
  // dist/chartjs-gauge.js
  // dist/chartjs-gauge.min.js
  {
    input,
    plugins: [
      babel({
        exclude: 'node_modules/**',
      }),
    ],
    output: {
      name: 'Gauge',
      file: 'dist/chartjs-gauge.js',
      banner,
      format: 'umd',
      indent: false,
      globals: {
        'chart.js': 'Chart',
      },
    },
    external: [
      'chart.js',
    ],
  },
  {
    input,
    plugins: [
      babel({
        exclude: 'node_modules/**',
      }),
      terser({
        output: {
          preamble: banner,
        },
      }),
    ],
    output: {
      name: 'Gauge',
      file: 'dist/chartjs-gauge.min.js',
      format: 'umd',
      indent: false,
      globals: {
        'chart.js': 'Chart',
      },
    },
    external: [
      'chart.js',
    ],
  },
];


src/controllers/controller.gauge.js
import Chart from 'chart.js';

// default options
Chart.defaults._set('gauge', {
  needle: {
    // Needle circle radius as the percentage of the chart area width
    radiusPercentage: 2,
    // Needle width as the percentage of the chart area width
    widthPercentage: 3.2,
    // Needle length as the percentage of the interval between inner radius (0%) and outer radius (100%) of the arc
    lengthPercentage: 80,
    // The color of the needle
    color: 'rgba(0, 0, 0, 1)',
  },
  valueLabel: {
    // fontSize: undefined
    display: true,
    formatter: null,
    color: 'rgba(255, 255, 255, 1)',
    backgroundColor: 'rgba(0, 0, 0, 1)',
    borderRadius: 5,
    padding: {
      top: 5,
      right: 5,
      bottom: 5,
      left: 5,
    },
    bottomMarginPercentage: 5,
  },
  animation: {
    duration: 1000,
    animateRotate: true,
    animateScale: false,
  },
  // The percentage of the chart that we cut out of the middle.
  cutoutPercentage: 50,
  // The rotation of the chart, where the first data arc begins.
  rotation: -Math.PI,
  // The total circumference of the chart.
  circumference: Math.PI,
  legend: {
    display: false,
  },
  tooltips: {
    enabled: false,
  },
});

const GaugeController = Chart.controllers.doughnut.extend({
  getValuePercent({ minValue, data }, value) {
    const min = minValue || 0;
    const max = data[data.length - 1] || 1;
    const length = max - min;
    const percent = (value - min) / length;
    return percent;
  },
  getWidth(chart) {
    return chart.chartArea.right - chart.chartArea.left;
  },
  getTranslation(chart) {
    const { chartArea, offsetX, offsetY } = chart;
    const centerX = (chartArea.left + chartArea.right) / 2;
    const centerY = (chartArea.top + chartArea.bottom) / 2;
    const dx = (centerX + offsetX);
    const dy = (centerY + offsetY);
    return { dx, dy };
  },
  getAngle({ chart, valuePercent }) {
    const { rotation, circumference } = chart.options;
    return rotation + (circumference * valuePercent);
  },
  /* TODO set min padding, not applied until chart.update() (also chartArea must have been set)
  setBottomPadding(chart) {
    const needleRadius = this.getNeedleRadius(chart);
    const padding = this.chart.config.options.layout.padding;
    if (needleRadius > padding.bottom) {
      padding.bottom = needleRadius;
      return true;
    }
    return false;
  },
  */
  drawNeedle(ease) {
    if (!this.chart.animating) { // triggered when hovering
      ease = 1;
    }
    const {
      ctx,
      config,
      innerRadius,
      outerRadius,
    } = this.chart;
    const dataset = config.data.datasets[this.index];
    const { previous } = this.getMeta();
    const {
      radiusPercentage,
      widthPercentage,
      lengthPercentage,
      color,
    } = config.options.needle;

    const width = this.getWidth(this.chart);
    const needleRadius = (radiusPercentage / 100) * width;
    const needleWidth = (widthPercentage / 100) * width;
    const needleLength = (lengthPercentage / 100) * (outerRadius - innerRadius) + innerRadius;

    // center
    const { dx, dy } = this.getTranslation(this.chart);

    // interpolate
    const origin = this.getAngle({ chart: this.chart, valuePercent: previous.valuePercent });
    // TODO valuePercent is in current.valuePercent also
    const target = this.getAngle({ chart: this.chart, valuePercent: this.getValuePercent(dataset, dataset.value) });
    const angle = origin + (target - origin) * ease;

    // draw
    ctx.save();
    ctx.translate(dx, dy);
    ctx.rotate(angle);
    ctx.fillStyle = color;

    // draw circle
    ctx.beginPath();
    ctx.ellipse(0, 0, needleRadius, needleRadius, 0, 0, 2 * Math.PI);
    ctx.fill();

    // draw needle
    ctx.beginPath();
    ctx.moveTo(0, needleWidth / 2);
    ctx.lineTo(needleLength, 0);
    ctx.lineTo(0, -needleWidth / 2);
    ctx.fill();

    ctx.restore();
  },
  drawValueLabel(ease) { // eslint-disable-line no-unused-vars
    if (!this.chart.config.options.valueLabel.display) {
      return;
    }
    const { ctx, config } = this.chart;
    const {
      defaultFontFamily,
    } = config.options;
    const dataset = config.data.datasets[this.index];
    const {
      formatter,
      fontSize,
      color,
      backgroundColor,
      borderRadius,
      padding,
      bottomMarginPercentage,
    } = config.options.valueLabel;

    const width = this.getWidth(this.chart);
    const bottomMargin = (bottomMarginPercentage / 100) * width;

    const fmt = formatter || (value => value);
    const valueText = fmt(dataset.value).toString();
    ctx.textBaseline = 'middle';
    ctx.textAlign = 'center';
    if (fontSize) {
      ctx.font = `${fontSize}px ${defaultFontFamily}`;
    }

    // const { width: textWidth, actualBoundingBoxAscent, actualBoundingBoxDescent } = ctx.measureText(valueText);
    // const textHeight = actualBoundingBoxAscent + actualBoundingBoxDescent;

    const { width: textWidth } = ctx.measureText(valueText);
    // approximate height until browsers support advanced TextMetrics
    const textHeight = Math.max(ctx.measureText('m').width, ctx.measureText('\uFF37').width);

    const x = -(padding.left + textWidth / 2);
    const y = -(padding.top + textHeight / 2);
    const w = (padding.left + textWidth + padding.right);
    const h = (padding.top + textHeight + padding.bottom);

    // center
    let { dx, dy } = this.getTranslation(this.chart);
    // add rotation
    const rotation = this.chart.options.rotation % (Math.PI * 2.0);
    dx += bottomMargin * Math.cos(rotation + Math.PI / 2);
    dy += bottomMargin * Math.sin(rotation + Math.PI / 2);

    // draw
    ctx.save();
    ctx.translate(dx, dy);

    // draw background
    ctx.beginPath();
    Chart.helpers.canvas.roundedRect(ctx, x, y, w, h, borderRadius);
    ctx.fillStyle = backgroundColor;
    ctx.fill();

    // draw value text
    ctx.fillStyle = color || config.options.defaultFontColor;
    const magicNumber = 0.075; // manual testing
    ctx.fillText(valueText, 0, textHeight * magicNumber);

    ctx.restore();
  },
  // overrides
  update(reset) {
    const dataset = this.chart.config.data.datasets[this.index];
    dataset.minValue = dataset.minValue || 0;

    const meta = this.getMeta();
    const initialValue = {
      valuePercent: 0,
    };
    // animations on will call update(reset) before update()
    if (reset) {
      meta.previous = null;
      meta.current = initialValue;
    } else {
      dataset.data.sort((a, b) => a - b);
      meta.previous = meta.current || initialValue;
      meta.current = {
        valuePercent: this.getValuePercent(dataset, dataset.value),
      };
    }
    Chart.controllers.doughnut.prototype.update.call(this, reset);
  },
  updateElement(arc, index, reset) {
    // TODO handle reset and options.animation
    Chart.controllers.doughnut.prototype.updateElement.call(this, arc, index, reset);
    const dataset = this.getDataset();
    const { data } = dataset;
    // const { options } = this.chart.config;
    // scale data
    const previousValue = index === 0 ? dataset.minValue : data[index - 1];
    const value = data[index];
    const startAngle = this.getAngle({ chart: this.chart, valuePercent: this.getValuePercent(dataset, previousValue) });
    const endAngle = this.getAngle({ chart: this.chart, valuePercent: this.getValuePercent(dataset, value) });
    const circumference = endAngle - startAngle;

    arc._model = {
      ...arc._model,
      startAngle,
      endAngle,
      circumference,
    };
  },
  draw(ease) {
    Chart.controllers.doughnut.prototype.draw.call(this, ease);

    this.drawNeedle(ease);
    this.drawValueLabel(ease);
  },
});

export default GaugeController;
